core/settings/qPrefDiveComputer.h

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: GPL-2.0
+#ifndef QPREFSDIVECOMPUTER_H
+#define QPREFSDIVECOMPUTER_H
+#include "core/pref.h"
+
+#include <QObject>
+
+class qPrefDiveComputer : public QObject {
+	Q_OBJECT
+	Q_PROPERTY(QString device READ device WRITE set_device NOTIFY device_changed);
+	Q_PROPERTY(QString device_name READ device_name WRITE set_device_name NOTIFY device_name_changed);
+	Q_PROPERTY(int download_mode READ download_mode WRITE set_download_mode NOTIFY download_mode_changed);
+	Q_PROPERTY(QString product READ product WRITE set_product NOTIFY product_changed);
+	Q_PROPERTY(QString vendor READ vendor WRITE	set_vendor NOTIFY vendor_changed);
+
+public:
+	qPrefDiveComputer(QObject *parent = NULL);
+	static qPrefDiveComputer *instance();
+
+	// Load/Sync local settings (disk) and struct preference
+	void loadSync(bool doSync);
+	void load() {loadSync(false); }
+	void sync() {loadSync(true); }
+
+public:
+	static inline const QString device() {return QString(prefs.dive_computer.device); };
```

Comment:
`inline` is redundant here.
When I read a class definition I'm not particularly interested in implementation details, so I think @dirkhh has a point.

Reply:
actuslly it is not. This instruct the compiler to use inline where possible (see above) but leave 1 static copy of the function for the linker.

If I remove inline, we will not benefit from the normal compiler optimization.

But I have no special heart here, if the consensus is not to use what clang offers.

Reply:
You are completely wrong on this one.

Defining a function in a class makes it inline, period. See first paragraph here: https://en.cppreference.com/w/cpp/language/inline
> A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.

It *must* be this way, otherwise your program would fail to llink. I think you misunderstood what the `inline` specifier does: It means that you can define the same function in multiple-translation units without the linker complaining. The compiler is free to choose whether it actually "inlines" the function.

So the `inline` here is 100% redundant. If clang interprets a redundant `inline` as forceinline or alwaysinline attribute (could you please provide a link for that?), then this is voodoo and we should not poison the code-base with such compiler-idiosyncrasies. If you insist on forcing inline, please do it in a compiler independent way using the `forceinline` and `always_inline` attributes.

Reply:
Please I am not completely wrong, it is and was inline (allowing the compiler to choose), however specifying it does not make it wrong ! @dirkhh was in doubt, the word “inline” albeit not needed, makes it clear to everyone.

I was perfectly happy with the old code, but the comments from you and @dirkhh called for a change.

Reply:
You really do not need to explain compiler basics, and I do not want to fight a war over idiology, C++ allows many things that are not 100% needed, but are nice to avoid confusion by people who are less C++ gurus.

I will leave it to @dirkhh to decide what he wants with this PR. It is really a bad motivation, to tell me “I am completely wrong”, when I try hard to react positive to changes requested.

Reply:
> I was perfectly happy with the old code, but the comments from you and @dirkhh called for a change.

Story of my life... :-/

Reply:
Note that the "completely wrong" was with respect to the claim that removing the `inline` keyword disables compiler optimization. I stand by that, even though I could have worded it less strongly. I think you take things too personal, there's no offense intended from my side whatsoever.

Reply:
Well the sentence “you are completely wrong” is kind of personal, and if you read my comment I explained why “static inline” makes sense, not why load() should have the inline keyword.

Reply:
Sorry, but this is not true. The argument went like this (verbatim quotes):
>> inline is redundant here.

> actuslly it is not. [...] If I remove inline, we will not benefit from the normal compiler optimization.

That's plain and simply wrong; sorry for being so blunt about this. This is neither an ideology nor a style question, simply a fact. And the whole thread proves that someone *does* have to explain language basics. I honestly fail to see the problem with that.

PS: I am very far from being a C++ expert, much less a guru.

Reply:
_**SIGH**_
We are all working on the same project, in our spare time, because it's fun.
No one is trying to attack anyone else. No one should feel attacked.
Almost all of us are using their second, or third, or fourth language for communication.

I really wish I wasn't starting to dread opening the discussion threads on PRs - but lately that's exactly what I've been feeling. And no, this is not a subtle, backwards way to criticize any one individual, either.

The purpose of the discussions on PRs is to make code better, to make sure your stupid, under-educated maintainer actually understands what's going on, and that overall the project is going in the right direction.

That can and does include disagreements. But those disagreements by definition are about the facts at stake. "you are wrong" sounds personal, but it's actually still about the facts at stake. When Linus tells me "you are an idiot" he is (a) correct and (b) crossing the line towards attacking the person (and he has had my explicit consent to that for 25+ years).

I cannot tell you how much I appreciate @janiversen 's work on cleaning up the C++/QML interface, much of which is broken because I wrote it and had no idea what I was doing. And I cannot tell you how much I appreciate that @bstoeger helps explain things to me that are outside my comfort zone. Because I honestly didn't know the subtleties of const, static, inline, references and what not in these class definitions and declarations. I love it when he corrects my misconceptions. And I appreciate it when he responds to my questions.

Does this mean that I'm potentially the wrong maintainer and should pass on the baton to someone else? I don't know. You tell me.

But none of this is picking on anyone, none of this is personal, none of this is rude or offensive.
It's intended to make the code better.

I don't want to lose the fun that is maintaining Subsurface (the assholes insulting me on FB for the shit software that I apparently force them to look at are bad enough, TBH). So let's all please remind ourselves why we do this and what our goal is. A better C++/QML interface that allows us to work on cool features for Subsurface-mobile (and Subsurface).

I'll get off my soap box. It feels like I said all of this before.